### PR TITLE
Make the SSL_COMP_add_compression_method a no-op instead of fixing it

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Note: current implementation is bogus and moreover it is impossible by
design to create custom compression methods in application code,
because the structure type COMP_METHOD is now opaque.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
Is it better to do this, or to fix things so that creating a custom compression method is possible? I can't think of a good reason why any new application would need one, but maybe legacy apps seeking to upgrade to 1.1.0 might need to - and this could represent a barrier to that.

**Comment by mattcaswell** [comment]
@vdukhovni I'd be interested in your view on this.

**Comment by levitte** [comment]
Additionally, I think that this is a breach of our version policy as well.  Granted, it was practically deprecated anyway, but that's a bug that needs fixing *so it works as documented / intended* rather than deprecating it for real.

**Comment by bernd-edlinger** [comment]
All I can say, is that it feels kind of pointless to invent new API functions for something
like that.

If you are concerned about legacy apps, there should be a legacy configuration
mode, that allows a legacy app without any change, i.e. make the internal stuff available
if configured for legacy apps, and just make sure that the installed OpenSSL version
is identical to the version of the header files.

**Comment by levitte** [comment]
We have a mechanism like that, but that works the other way around.  `./config no-deprecated` will only give you the most current interface and nothing more (i.e. not functions marked for deprecation, which you can find if you search for `OPENSSL_API_COMPAT` and `DEPRECATEDIN` in the public headers.  The default build includes the deprecated APIs.

**Comment by richsalz** [comment]
Compression at the TLS layer is stupid and insecure.  It just is.  Nevertheless, if we wanted to support adding new compression schemes at runtime, then a single new API that took the callback functions and returns a ```COMP_method``` seems the way to go.

**Comment by richsalz** [comment]
It is a bug.  It's not worth fixing.  See #2615.  Right now you can't call the API, let's leave it as an open bug.

**Comment by vdukhovni** [comment]
My take is that given that 1.1.0 has no support for compression methods, and that no new compression codepoints in TLS are in our future, the fix is to unconditionally return an error (not success) for any attempt to add a custom compression method. Additionally, on platforms that support compile-time warnings, we should emit a warning when compiling code that uses this interface. We could even update the headers to redirect this function to a non-existent name, so that linking fails, while retaining the implementation to avoid run-time relocation errors in already compiled code.

Any applications that used custom compression methods (all zero of them I expect) porting to 1.1.0 will then stop calling the no longer supported API.

Bottom line, in this case I'm not a stickler for 1.0.x compatibility, and we're deprecating this, we should do it in as noisy a way as possible, so that there are no latent runtime surprises.

**Comment by bernd-edlinger** [comment]
I tried to replace a dangerous bug with something harmless:

That's what I mean:
```
diff --git a/test/ssltest_old.c b/test/ssltest_old.c
index 1bfa983..da72a3a 100644
--- a/test/ssltest_old.c
+++ b/test/ssltest_old.c
@@ -1406,7 +1406,7 @@ int main(int argc, char *argv[])
 /*      if (cipher == NULL) cipher=getenv("SSL_CIPHER"); */
 
 #ifndef OPENSSL_NO_COMP
-    if (comp == COMP_ZLIB)
+        comp = 222;
         cm = COMP_zlib();
     if (cm != NULL) {
         if (COMP_get_type(cm) != NID_undef) {
```

```
./config zlib && make
make test => OK
MALLOC_PERTURB_=123 make test => FAIL:
../test/recipes/80-test_ssl_old.t .......... 1/7     
    #   Failed test 'test sslv2/sslv3 via BIO pair'
    #   at ../test/recipes/80-test_ssl_old.t line 354.
    
    #   Failed test 'test dtlsv1'
    #   at ../test/recipes/80-test_ssl_old.t line 362.
    
    #   Failed test 'test dtlsv1 with server authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 364.
    
    #   Failed test 'test dtlsv1 with client authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 366.
    
    #   Failed test 'test dtlsv1 with both server and client authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 368.
    
    #   Failed test 'test dtlsv1.2'
    #   at ../test/recipes/80-test_ssl_old.t line 376.
    
    #   Failed test 'test dtlsv1.2 with server authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 378.
    
    #   Failed test 'test dtlsv1.2 with client authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 380.
    
    #   Failed test 'test dtlsv1.2 with both server and client authentication'
    #   at ../test/recipes/80-test_ssl_old.t line 382.
    
    #   Failed test 'test sslv2/sslv3 w/o (EC)DHE via BIO pair'
    #   at ../test/recipes/80-test_ssl_old.t line 393.
```
